pkg/logs/sender/batch_strategy.go

```diff
@@ -89,14 +100,25 @@ func (s *batchStrategy) Send(inputChan chan *message.Message, outputChan chan *m
 		flushTicker.Stop()
 		s.pendingSends.Wait()
 	}()
+	var startIdle = time.Now()
 	for {
 		select {
 		case m, isOpen := <-inputChan:
+
 			if !isOpen {
 				// inputChan has been closed, no more payloads are expected
 				return
 			}
+			s.senderIdleTimeMs.Add(float64(time.Since(startIdle) / time.Millisecond))
+			var startInUse = time.Now()
+
 			s.processMessage(m, outputChan, send)
+
+			s.senderInUseTimeMs.Add(float64(time.Since(startInUse) / time.Millisecond))
+			startIdle = time.Now()
+
+			batchStrategyExpVars.Set(s.getExpVarName(), expvar.Func(s.publishExpVars))
```

Comment:
Let me give that a try. TBH i wasn't sure what the "correct" way was to make a nested expvar structure like this. 
But my original rationale behind keeping the float expvars on the struct was getting the atomic increment for free, but I suppose the expvar map comes with it's own lock for that. 

RE: performance - ive ran some benchmarks (granted not massive stress tests) but it doesn't not noticeably impact "high throughput" ~6k MB/sec

Reply:
> expvar map comes with it's own lock for that


You can keep the float expvars for that exact purpose but you can replace the `expvar.Func` 1:1 with `expvar.Map` to avoid higher number of type conversions (the runner expvars has some usage like that [here](https://github.com/DataDog/datadog-agent/blob/d9363eed060c1af79b0b4c21520c05a1373076c4/pkg/collector/runner/expvars/workers.go#L45-L47)).


> RE: performance


For the loop part, I think you can probably init most of the expvar structure outside of the message loop with some placeholder values and then just set the `sender*TimeMs` vars within the loop itself. If it's not that big of an impact on performance, it may not matter though so it's up to you.

Reply:
💯